consensus/istanbul/core/handler_test.go

```diff
@@ -464,6 +483,236 @@ func TestCore_handlerMsg(t *testing.T) {
+		}
+	}
+	return subCN, remainingCN
+}
+
+// Simulate a proposer that receives messages from disagreeing groups of CNs.
+func simulateMaliciousCN(t *testing.T, numValidators int, numMalicious int) State {
+	if testing.Verbose() {
+		enableLog()
+	}
+
+	fork.SetHardForkBlockNumberConfig(&params.ChainConfig{})
+	defer fork.ClearHardForkBlockNumberConfig()
+
+	// Note that genValidators(n) will generate n/3 validators.
+	// We want n validators, thus calling genValidators(3n).
+	validatorAddrs, validatorKeyMap := genValidators(numValidators * 3)
+
+	// Add more EXPECT()s to remove unexpected call error
+	mockBackend, mockCtrl := newMockBackend(t, validatorAddrs)
+	mockBackend.EXPECT().Commit(gomock.Any(), gomock.Any()).Return(nil).AnyTimes()
+	mockBackend.EXPECT().HasBadProposal(gomock.Any()).Return(true).AnyTimes()
+	defer mockCtrl.Finish()
+
+	var (
+		// it creates two pre-defined blocks: one for benign CNs, the other for the malicious
+		// newProposal is a block which the proposer has created
+		// malProposal is an incorrect block that malicious CNs use to try stop consensus
+		lastProposal, _ = mockBackend.LastProposal()
+		lastBlock       = lastProposal.(*types.Block)
+		validators      = mockBackend.Validators(lastBlock)
+		proposer        = validators.GetProposer()
+		proposerKey     = validatorKeyMap[proposer.Address()]
+		newProposal, _  = genBlockParams(lastBlock, proposerKey, 0, 1, 1)
+		malProposal, _  = genBlockParams(lastBlock, proposerKey, 0, 0, 0)
+	)
+
+	// Start istanbul core
+	istConfig := istanbul.DefaultConfig
+	istConfig.ProposerPolicy = istanbul.WeightedRandom
+	istCore := New(mockBackend, istConfig).(*core)
+	err := istCore.Start()
+	require.Nil(t, err)
+	defer istCore.Stop()
+
+	// Step 1 - Pre-prepare with correct message
+
+	// Create pre-prepare message
+	istanbulMsg, err := genIstanbulMsg(msgPreprepare, lastBlock.Hash(), newProposal, proposer.Address(), proposerKey)
+	require.Nil(t, err)
+
+	// Handle pre-prepare message
+	err = istCore.handleMsg(istanbulMsg.Payload)
+	require.Nil(t, err)
+
+	// splitSubList split current committee into benign CNs and malicious CNs
+	subList := validators.SubList(lastBlock.Hash(), istCore.currentView())
+	maliciousCNs, benignCNs := splitSubList(subList, numMalicious, proposer.Address())
+	benignCNs = append(benignCNs, proposer)
+
+	// Shortcut for sending consensus message to everyone in `CNList`
+	sendMessages := func(state uint64, proposal *types.Block, CNList []istanbul.Validator) {
+		for _, val := range CNList {
+			valKey := validatorKeyMap[val.Address()]
+			istanbulMsg, err := genIstanbulMsg(state, lastBlock.Hash(), proposal, val.Address(), valKey)
+			assert.Nil(t, err)
+			err = istCore.handleMsg(istanbulMsg.Payload)
+			// assert.Nil(t, err)
+		}
+	}
+
+	// Step 2 - Receive disagreeing prepare messages
+
+	sendMessages(msgPrepare, newProposal, benignCNs)
+	sendMessages(msgPrepare, malProposal, maliciousCNs)
+
+	if istCore.state.Cmp(StatePrepared) != 0 {
```

Comment:
```suggestion
	if istCore.state.Cmp(StatePreprepared) == 0 {
```
Isn't this more accurate? (It will be checked at the caller though)

Reply:
I agree with your opinion. reflected at [2631583](https://github.com/klaytn/klaytn/pull/1396/commits/263158306e17600f61feab25f09d487083324cbc).